#! /bin/sh /usr/share/dpatch/dpatch-run
## fix_CreateDelegate_ArgumentException.dpatch by Mirco Bauer <meebey@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch taken from upstream SVN revision 130175
## DP: 2009-03-25  Jb Evain  <jbevain@novell.com>
## DP: * Delegate.cs (CreateDelegate): fix checks to deal with
## DP:   valuetypes -> obj conversions for arguments, and avoid
## DP:   issues with such unallowed conversion for return types.

@DPATCH@
diff -urNad git~/mcs/class/corlib/System/Delegate.cs git/mcs/class/corlib/System/Delegate.cs
--- git~/mcs/class/corlib/System/Delegate.cs	2009-07-15 20:52:46.000000000 +0200
+++ git/mcs/class/corlib/System/Delegate.cs	2009-09-07 21:55:10.000000000 +0200
@@ -133,7 +133,7 @@
 
 			// Delegate contravariance
 			if (!match) {
-				if (!delArgType.IsValueType && (delArgType != typeof (ValueType)) && (argType.IsAssignableFrom (delArgType)))
+				if (!argType.IsValueType && argType.IsAssignableFrom (delArgType))
 					match = true;
 			}
 
@@ -147,7 +147,7 @@
 #if NET_2_0
 			if (!returnMatch) {
 				// Delegate covariance
-				if (!delReturnType.IsValueType && (delReturnType != typeof (ValueType)) && (delReturnType.IsAssignableFrom (returnType)))
+				if (!returnType.IsValueType && delReturnType.IsAssignableFrom (returnType))
 					returnMatch = true;
 			}
 #endif
